crates/rattler_installs_packages/src/venv.rs

```diff
@@ -0,0 +1,177 @@
+pub enum PythonLocation {
+    /// Use system interpreter
+    System,
+    // Use custom interpreter
+    Custom(PathBuf),
+}
+
+impl PythonLocation {
+    /// Location of python executable
+    pub fn executable(&self) -> Result<PathBuf, FindPythonError> {
+        match self {
+            PythonLocation::System => system_python_executable(),
+            PythonLocation::Custom(path) => Ok(path.clone()),
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum VEnvError {
+    #[error(transparent)]
+    FindPythonError(#[from] FindPythonError),
+    #[error(transparent)]
+    ParsePythonInterpreterVersionError(#[from] ParsePythonInterpreterVersionError),
+    #[error("failed to run 'python -m venv': `{0}`")]
+    FailedToRun(String),
+    #[error(transparent)]
+    FailedToCreate(#[from] std::io::Error),
+}
+
+/// Represents a virtual environment in which wheels can be installed
+pub struct VEnv {
+    /// Location of the virtual environment
+    location: PathBuf,
+    /// Install paths for this virtual environment
+    install_paths: InstallPaths,
+    /// Specifies if this is a windows venv
+    windows: bool,
+}
+
+impl VEnv {
+    fn new(location: PathBuf, install_paths: InstallPaths, windows: bool) -> Self {
+        Self {
+            location,
+            install_paths,
+            windows,
+        }
+    }
+
+    /// Install a wheel into this virtual environment
+    pub fn install_wheel(
+        &self,
+        wheel: &Wheel,
+        options: &UnpackWheelOptions,
+    ) -> Result<(), UnpackError> {
+        wheel.unpack(&self.location, &self.install_paths, options)
+    }
+
+    /// Path to binaries in venv
+    pub fn binaries(&self) -> PathBuf {
+        if self.windows {
+            self.location.join("Scripts")
+        } else {
+            self.location.join("bin")
+        }
+    }
+
+    /// Path to python executable in venv
+    pub fn python_executable(&self) -> PathBuf {
+        if self.windows {
+            self.binaries().join("python.exe")
+        } else {
+            self.binaries().join("python")
+        }
+    }
+
+    /// Execute python script in venv
+    pub fn execute_script(&self, source: &Path) -> std::io::Result<Output> {
+        let mut cmd = Command::new(self.python_executable());
+        cmd.arg(source);
+        cmd.output()
+    }
+
+    /// Execute python command in venv
+    pub fn execute_command<S: AsRef<str>>(&self, command: S) -> std::io::Result<Output> {
+        let mut cmd = Command::new(self.python_executable());
+        cmd.arg("-c");
+        cmd.arg(command.as_ref());
+        cmd.output()
+    }
+}
+
+/// Create a virtual environment at specified directory
+/// for the platform we are running on
+pub fn venv(venv_dir: &Path, python: PythonLocation) -> Result<VEnv, VEnvError> {
+    custom_venv(venv_dir, python, cfg!(windows))
+}
+
+/// Create a virtual environment at specified directory
+/// allows specifying if this is a windows venv
+pub fn custom_venv(
```

Comment:
Would it make sense to make this an associate function on `VEnv`?

And I would add the `create` verb to the name of the function.

Reply:
Yeah I was kinda following along the lines of how tempfolder does it. But I don't really mind either way.